<a href="https://colab.research.google.com/github/binteaamer/FlyrankInternship/blob/main/work/notebooks/w01_research_question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/binteaamer/FlyrankInternship/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

In [2]:
!git clone https://github.com/binteaamer/FlyrankInternship.git
%cd FlyrankInternship

Cloning into 'FlyrankInternship'...
remote: Enumerating objects: 152, done.
remote: Counting objects: 100% (152/152), done.
remote: Compressing objects: 100% (108/108), done.
remote: Total 152 (delta 60), reused 92 (delta 28), pack-reused 0 (from 0)
Receiving objects: 100% (152/152), 1.85 MiB | 4.79 MiB/s, done.
Resolving deltas: 100% (60/60), done.
/content/FlyrankInternship


In [3]:
import pandas as pd

df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

print("Shape:", df.shape)

print("\nColumns:")
print(df.columns.tolist())

print("\nData types:")
print(df.dtypes)

print("\nMissing values:")
print(df.isna().sum().sort_values(ascending=False).head(20))

Shape: (30000, 44)

Columns:
['content_id', 'client_id', 'search_volume', 'competition', 'competition_level', 'cpc', 'content_type', 'main_intent', 'word_count', 'char_count', 'provider_used', 'model_used', 'impressions_90d', 'clicks_90d', 'pageviews_90d', 'sessions_90d', 'users_90d', 'engaged_sessions_90d', 'ai_sessions_90d', 'scroll_events_90d', 'days_with_impressions', 'days_with_sessions', 'impressions_last_30d', 'clicks_last_30d', 'sessions_last_30d', 'impressions_prev_30d', 'clicks_prev_30d', 'sessions_prev_30d', 'content_age_days', 'age_tier', 'age_tier_order', 'days_since_last_update', 'freshness_tier', 'word_count_tier', 'char_count_tier', 'ctr', 'avg_position', 'engagement_rate', 'scroll_rate', 'ai_traffic_pct', 'impression_tier', 'position_tier', 'trend_direction', 'trend_pct']

Data types:
content_id                 object
client_id                  object
search_volume             float64
competition               float64
competition_level          object
cpc              

## 1. My lane (or freestyle) and why

**Freestyle — Quick-Win Opportunity Detection**

I want to investigate which pages are closest to a meaningful ranking improvement and should be prioritized for targeted SEO optimization.

The idea is to focus on pages that already have search visibility but are not yet in the strongest ranking positions. These pages may have an opportunity for improvement through actions such as reviewing the title, meta description, search intent, or on-page content.

I chose this freestyle lane because the starter dataset contains page-level search position, impressions, clicks, CTR, content characteristics, and engagement signals. These signals can potentially be combined to identify pages with both meaningful search demand and room for ranking improvement.

The longer-term goal is to use historical performance data to determine whether pages identified as quick-win opportunities are more likely to achieve a meaningful ranking improvement in a future period.

This makes the project different from simply finding declining pages. Instead of asking which pages need to be fixed after performance has declined, the project asks which visible pages may be worth proactively optimizing.



In [4]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 2. The question: decision, action, cost of a wrong call

### Decision

Which pages should be prioritized as potential quick-win opportunities for targeted SEO optimization?

### Who acts?

An SEO specialist or content strategist reviews the prioritized pages and decides whether an optimization is appropriate.

### Action

For a page identified as a potential quick win, the specialist could:

- review the page's search position and impressions;
- review the title and meta description;
- check whether the content matches the search intent;
- review the page's content and structure;
- make a targeted SEO improvement if appropriate;
- monitor the page's ranking after the change.

The model would support prioritization. It would not automatically decide which SEO change should be made.

### Cost of a wrong recommendation

A false positive could cause the SEO team to spend time optimizing a page that has little potential for improvement.

A false negative could cause the team to miss a page that had a good opportunity for a ranking improvement.

The cost is therefore mainly wasted optimization effort or missed opportunities.

### Why data or ML can help

A page's average search position alone does not tell us whether it is a good quick-win opportunity.

Two pages can have similar positions but very different impressions, CTR, content characteristics, and engagement.

ML may be useful because it can combine several signals and learn patterns associated with pages that later achieve meaningful ranking improvements.

The goal is therefore not simply to produce a score. The goal is to create a ranked list that helps an SEO specialist decide where to spend optimization effort first.

In [5]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 3. Quick look at the data

The starter dataset contains 30,000 content items and includes several signals that are relevant to the Quick-Win question.

In particular, it contains average search position, impressions, clicks, CTR, content age, freshness, engagement, and content characteristics.

These signals allow us to identify pages that already have measurable search visibility while still having room to improve their ranking.

The following code shows some basic numbers from the starter dataset.

In [6]:
import pandas as pd

df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

print("Total pages:", len(df))

print("Average search position:",
      round(df["avg_position"].mean(), 2))

print("Average impressions over 90 days:",
      round(df["impressions_90d"].mean(), 2))

print("Average CTR:",
      round(df["ctr"].mean(), 3))

print("\nPosition tier counts:")
print(df["position_tier"].value_counts())

Total pages: 30000
Average search position: 16.34
Average impressions over 90 days: 5200.37
Average CTR: 0.511

Position tier counts:
position_tier
page_1      11814
striking     7304
page_3_5     7242
top_3        2321
deep         1319
Name: count, dtype: int64


In [7]:
print("\nPages with average position between 5 and 15:")

quick_win_position = df[
    (df["avg_position"] >= 5) &
    (df["avg_position"] <= 15)
]

print(len(quick_win_position))


Pages with average position between 5 and 15:
13739


### What these numbers suggest

There are 30,000 pages in the starter dataset, with an average search position of 16.34.

I identified 13,739 pages with an average position between 5 and 15. This is approximately 45.8% of the dataset. These pages are visible in search but are not yet in the top 3 positions, making them a reasonable candidate group for investigating potential ranking opportunities.

However, being in positions 5–15 does not by itself mean that a page is a quick win. The next step is to determine whether other available signals, such as impressions, CTR, content characteristics, and engagement, can help distinguish pages with stronger future improvement potential.

Therefore, the 5–15 range is an initial candidate definition, not a proven rule.

## 4. Careful words: what I can and can't claim

At this stage, I can identify pages that have search visibility and are within a candidate ranking range for further investigation.

I cannot yet claim that these pages are definitely quick wins or that an SEO change will cause them to rank higher.

I also cannot claim that a page at position 5 is inherently a better opportunity than a page at position 15 without testing the relationship using historical data.

The starter dataset provides useful signals for identifying candidate pages, but the final prediction target should be based on a future performance window.

For the full project, I intend to investigate whether information available before an optimization opportunity can predict a meaningful future ranking improvement.

I will avoid claiming causation. A future ranking improvement would not necessarily prove that an SEO change caused the improvement, because rankings can change for many reasons.

The intended output is decision support: a ranked list of pages that appear worth reviewing first.

## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.